### PR TITLE
Add option to manully enter bearing reading.

### LINKED ISSUE #1572: Graceful degradation of bearing widget
#### Software and hardware versions 
Collect v1.10.2, 1.11.0, 
Android v5 (Lenovo Tab3 7"), Android 6 (Moto G4+), Android 7 (Samsung S2 8")

#### Problem description
The [bearing widget](http://docs.opendatakit.org/form-widgets/#bearing-widget) works great in devices with compass hardware (S2), but does not work in devices without the required hardware to measure magnetic bearings (Tab3, G4+).

#### Steps to reproduce the problem
Fill in a form with the bearing widget on a compatible vs. an incompatible device.

#### Expected behaviour
Incompatible devices show the "capture bearing" button greyed out, so there is some form of communication between ODK Collect rendering the button as inactive and the underlying hardware.

It would be great if in this case, the form would simply and gracefully degrade the widget to a simple numeric input -- after all, the bearing widget is only a special case of a numeric input.

This would allow devices to capture bearing information from built-in hardware (ideally) or from external sources (e.g. a hand held compass, if available).

### PR DESCRIPTION
**Pull request by akshay-ap**
Closes #1572
What has been done to verify that this works as intended?
Moto M - (with sensor support)
Why is this the best possible solution? Were any other approaches considered?
This method required minimum changes in current code. Changes are easy to understand.
I tried only addin `editText` and reomove `TextView` but some tests were failing. Might be due to dependence on `Textview`. So, had to keep both.

Are there any risks to merging this code? If so, what are they?
Some minor changes might be needed to make widget more user friendly. Like toasts with better messages.
isValidInput() method might need improvement.

Do we need any specific form for testing your changes? If so, please attach one.
Need to check if the changes affect form filling in any way and testing for Older devices needs to be done.

### CODE REVIEW

**Comment by akshay-ap** [comment]
Valid range is from 0 to 360.0.

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java [hunk 1]
```diff
@@ -147,9 +180,80 @@ private void checkForRequiredSensors() {
...
+        manualData.setFilters(fa);
+        manualData.setLayoutParams(params);
+
+        manualData.setOnEditorActionListener(new TextView.OnEditorActionListener() {
+            @Override
+            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+                String data = manualData.getText().toString();
+                if (!data.equals("") && isValidInput(data)) {
+                    answerDisplay.setText(manualData.getText().toString());
+                    ToastUtils.showShortToast("Data entered : " + data);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: I would rather not have these toasts! I think they are confusing. They also don't actually prevent users from entering invalid values. I don't think there should be any validation -- people who need some can add it in their forms instead.

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java [hunk 2]
```diff
@@ -73,29 +79,51 @@ public void onClick(View v) {
...
 
-        checkForRequiredSensors();
 
-        LinearLayout answerLayout = new LinearLayout(getContext());
-        answerLayout.setOrientation(LinearLayout.VERTICAL);
         answerLayout.addView(getBearingButton);
         answerLayout.addView(answerDisplay);
+        String s = prompt.getAnswerText();
+        if (s != null && !s.equals("")) {
+
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: This blank line shouldn't be here.

**Review by lognaturel — COMMENTED** [review]
My preference would still be for this to behave exactly like the external value widgets. That is, no validation and no separate echoing back of the value. It should be possible to use almost the same code to achieve this.

@grzesiek2010 what do you think?

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java (hunk 1 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: Got it. Will remove every type of validation.

**Comment by grzesiek2010** [comment]
> My preference would still be for this to behave exactly like the external value widgets. That is, no validation and no separate echoing back of the value. It should be possible to use almost the same code to achieve this.
@grzesiek2010 what do you think?

I agree with you.

**Comment by akshay-ap** [comment]
@lognaturel @grzesiek2010 This is how it looks now.

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java [hunk 3]
```diff
@@ -45,26 +50,25 @@
 @SuppressLint("ViewConstructor")
 public class BearingWidget extends QuestionWidget implements BinaryWidget {
     private Button getBearingButton;
-    private TextView answerDisplay;
+    private boolean isSensorAvailable = false;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: It's a false by default you don't need to initialize it.

**Comment by grzesiek2010** [comment]
It looks ok now.

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java (hunk 3 above)
Comment (akshay-ap) [reply collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: Should I push another commit?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: Yes please.

**Comment by grzesiek2010** [comment]
@opendatakit-bot label "changes requested"

**Comment by grzesiek2010** [comment]
@opendatakit-bot unlabel "reviewed"

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java [hunk 4]
```diff
@@ -73,34 +77,49 @@ public void onClick(View v) {
...
                                 getFormEntryPrompt().getIndex());
-                Intent i;
-                i = new Intent(getContext(), BearingActivity.class);
 
-                waitForData();
-                ((Activity) getContext()).startActivityForResult(i,
-                        RequestCodes.BEARING_CAPTURE);
+                if (isSensorAvailable) {
+                    Intent i;
+                    i = new Intent(getContext(), BearingActivity.class);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: You can do that in one line:
`Intent i = new Intent(getContext(), BearingActivity.class);`

I know it was in two lines before but you touched this fragment of code so it always good to use the [scout rule](http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule)

File: collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java [hunk 5]
```diff
@@ -73,34 +77,49 @@ public void onClick(View v) {
...
-        checkForRequiredSensors();
-
-        LinearLayout answerLayout = new LinearLayout(getContext());
-        answerLayout.setOrientation(LinearLayout.VERTICAL);
         answerLayout.addView(getBearingButton);
-        answerLayout.addView(answerDisplay);
+        answerLayout.addView(answer);
+        String s = prompt.getAnswerText();
+        if (s != null && !s.equals("")) {
+
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BearingWidget.java]: Please remove this empty line.

**Comment by akshay-ap** [comment]
@grzesiek2010 I have made the requested changes.

**Comment by grzesiek2010** [comment]
@akshay-ap 
Thanks.

@opendatakit-bot unlabel "changes requested"
@opendatakit-bot label "needs testing"

**Comment by kkrawczyk123** [comment]
### Tested with success!

Verified on Android: 5.1. (Huawei Y5) and 4.4 (Samsung Galaxy J1)

**Comment by kkrawczyk123** [comment]
@opendatakit-bot unlabel "needs testing"

**Comment by lognaturel** [comment]
@akshay-ap Once you fix the merge conflict this will be ready to get in!

**Comment by lognaturel** [comment]
@akshay-ap looks like there's a compilation error. You may want to look at https://github.com/opendatakit/collect/pull/1625 to understand the change that caused the conflict. If you need help fixing it, we can talk in Slack.

**Comment by akshay-ap** [comment]
@lognaturel I think I did some mistake while resolving conflict. Will look at it later.

**Comment by lognaturel** [comment]
Thanks, @akshay-ap! I finished the merge so we can get this in for the beta.

@mmarciniak90 @kkrawczyk123 I don't have a device with the sensors so please verify that case again in your release testing! 🙏

**Review by lognaturel — APPROVED** [review]

**Comment by mmarciniak90** [comment]
@lognaturel of course we will do it! This case is included in our test plan.

**Comment by lognaturel** [comment]
Of course, I should have known! 😊

**Comment by akshay-ap** [comment]
@lognaturel Thanks for fixing merge conflict !!!!

**Pull request merged**